File: opencv/template_matching/template_matching.py
import cv2

from rich import print
from rich.console import Console
console = Console()

console.print(rf" {'#'*20} TEMPLATE MATCHING WITH OPENCV {'#'*20} ", style="bold green")

def preprocess(image):
    if len(image.shape) == 3:
        gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    else:
        gray = image.copy()
    
    denoised = cv2.medianBlur(gray, 3)
    clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(8, 8))
    enhanced = clahe.apply(denoised)
    normalized = cv2.normalize(enhanced, None, 0, 256, cv2.NORM_MINMAX)
    
    return normalized    


search_image_name = input('Enter Path and Name Source Image(Search): ') 
template_image_name = input('Enter Path and Name Target Image(Template): ')
threshold = float(input('Enter Threshold(float): '))

search_image = cv2.imread(search_image_name)
template_image = cv2.imread(template_image_name, cv2.IMREAD_GRAYSCALE)

search_image_gray = search_image.copy()

search_image_gray = cv2.cvtColor(search_image_gray, cv2.COLOR_BGR2GRAY)

result = cv2.matchTemplate(search_image_gray, template_image, cv2.TM_CCOEFF_NORMED)

min_val, max_val, min_loc, max_loc = cv2.minMaxLoc(result)
print(f"Max Value: {max_val} - Max Loc: {max_loc} / Min Value: {min_val}")

# location = np.where(result>=threshold)
template_height, template_width = template_image.shape

# for pt in zip(*location[::-1]):
#     cv2.putText(search_image, f"Max_Value: {max_val:.2f}", max_loc, 1, 0.5, (255,0,0), 2,)
#     cv2.rectangle(search_image, pt, (pt[0]+template_width, pt[1]+template_height), (0,255,0), 2)

# Show Output
cv2.putText(search_image, f"Max_Value: {max_val:.2f}", (max_loc[0], max_loc[1]-5), 1, 1, (0,255,0), 2,)
cv2.rectangle(search_image, max_loc, (max_loc[0]+template_width, max_loc[1]+template_height), (0,255,0), 2)

cv2.imwrite("./outputs/python_output.jpg", search_image)
cv2.namedWindow("Template Matching Output", cv2.WINDOW_FREERATIO)
cv2.imshow("Template Matching Output", search_image)

# Stop the program on the ESC key or 'q'
keyCode = cv2.waitKey(0) & 0xFF
if keyCode == 27 or keyCode == ord('q'):
    cv2.destroyAllWindows()
    

File: opencv/template_matching/template_matching.cpp
#include <iostream>
#include <opencv2/opencv.hpp>

using namespace std;
using namespace cv;


int main(){
    cv::Mat template_image = cv::imread("./images/template.png", cv::IMREAD_GRAYSCALE);
    cv::Mat search_image = cv::imread("./images/search.png");

    cv::Mat search_image_gray;
    cv::cvtColor(search_image, search_image_gray, cv::COLOR_BGR2GRAY);


    if (template_image.empty()) {
        std::cerr << "Failed to load template_image\n";
        return -1;
    }
    if (search_image.empty()) {
        std::cerr << "Failed to load search_image\n";
        return -1;
    }
    
    cv::Mat result;
    
    cv::matchTemplate(search_image_gray, template_image, result, cv::TM_CCOEFF_NORMED);

    double min_val, max_val;
    cv::Point min_loc, max_loc;
    cv::minMaxLoc(result, &min_val, &max_val, &min_loc, &max_loc);

    int template_width = template_image.cols;
    int template_height = template_image.rows;

    cv::Point pt2(max_loc.x + template_width, max_loc.y + template_height);

    cv::putText(search_image, "Max_Value" + std::to_string(max_val), cv::Point(max_loc.x, max_loc.y-5), 1, 1, cv::Scalar(0,255,0), 2);
    cv::rectangle(search_image, max_loc, pt2, cv::Scalar(0,255,0), 2);
    
    
    cv::imwrite("./outputs/cpp_output.jpg", search_image);
    cv::imshow("Outpout", search_image);
    cv::waitKey(0);

    return 0;
}